src/shogun/mathematics/linalg/internal/implementation/VectorMean.h

```diff
@@ -0,0 +1,134 @@
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ */
+
+#ifndef VECTOR_MEAN_IMPL_H_
+#define VECTOR_MEAN_IMPL_H_
+
+#include <shogun/lib/config.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/mathematics/linalg/internal/implementation/VectorSum.h>
+
+#ifdef HAVE_VIENNACL
+#include <shogun/lib/GPUVector.h>
+#include <shogun/lib/GPUMatrix.h>
+#endif
+
+#include <shogun/mathematics/eigen3.h>
+
+#include <numeric>
+
+namespace shogun
+{
+
+namespace linalg
+{
+
+/**
+ * All backend specific implementations are defined within this namespace
+ */
+namespace implementation
+{
+
+/**
+ * @brief Generic class vector_mean which provides a static compute method. This class
+ * is specialized for different types of vectors and backend, providing a mean
+ * to deal with various vectors directly without having to convert
+ */
+template <enum Backend, class Vector>
+struct vector_mean
+{
+	/** Scalar type */
+	typedef typename Vector::Scalar T;
+
+	/**
+	 * Method that computes the vector mean
+	 *
+	 * @param a vector whose mean has to be computed
+	 * @return the vector mean \f$\mean_i a_i\f$
+	 */
+	static T compute(Vector a);
+};
+
+
+/**
+ * @brief Specialization of generic vector_mean for the Eigen3 backend
+ */
+template <class Vector>
+struct vector_mean<Backend::EIGEN3, Vector>
+{
+	/** Scalar type */
+	typedef typename Vector::Scalar T;
+
+	/**
+	 * Method that computes the mean of SGVectors using Eigen3
+	 *
+	 * @param vec a vector whose mean has to be computed
+	 * @return the vector mean \f$\mean_i a_i\f$
+	 */
+	static T compute(SGVector<T> vec)
+	{
+		REQUIRE(vec.vlen > 0, "Vector can not be empty!\n");
+                return (vector_sum<Backend::EIGEN3, SGVector<T> >::compute(vec)
```

Comment:
@OXPHOS yes it would look lame. What I was thinking that - say, have a trait.
```c++
template <typename T>
struct some_trait
{
    using Scalar = T;
};
```
and then specialize it as 
```c++
template <>
struct some_trait<int32_t> // similarly for other int types
{
    using Scalar = float64_t;
};
```
so your method signature becomes
```c++
template <class T>
typename some_trait<typename T::Scalar>::Scalar mean(T op);
// T = SGVector or SGMatrix, both have a Scalar defined
```
Yes we are not using `decltype` for this one. But this is one way we can achieve this - no ugly `if-else`. It still solves the return type problem for *all* the types without losing precision. Let me know if there are better/other ways you can think of to handle this.

I am not sure whether this is the way to go though. Say, if I have a int vector with 1s and 0s, as long as there is a single 0, I'd get 0 mean every time if the return type is just int and not some floating type (which would be the case if we didn't do type-cast while `/`-ing). Is that the expected behavior? Not sure, just thinking aloud.